src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+                        filterText.Length <= 1)
+                    {
+                        initialListOfItemsToBeIncluded.Add(new ExtendedFilterResult(item, new FilterResult(roslynItem, filterText, matchedFilterText: false)));
+                    }
+                }
+            }
+
+            // If the session was created/maintained out of Roslyn, e.g. in debugger; no properties are set and we should use data.Snapshot.
+            // However, we prefer using the original snapshot in some projection scenarios.
+            var snapshotForDocument = data.InitialSortedList
+                                          .FirstOrDefault(i => i.Properties.ContainsProperty(CompletionSource.TriggerSnapshot))?
+                                          .Properties.GetProperty<ITextSnapshot>(CompletionSource.TriggerSnapshot)
+                                          ?? data.Snapshot;
+
+            var document = snapshotForDocument.GetOpenDocumentInCurrentContextWithChanges();
+            var completionService = document?.GetLanguageService<CompletionService>();
+            var completionRules = completionService?.GetRules() ?? CompletionRules.Default;
+
+            if (data.Trigger.Reason == AsyncCompletionData.CompletionTriggerReason.Backspace &&
+                completionRules.DismissIfLastCharacterDeleted &&
+                session.ApplicableToSpan.GetText(data.Snapshot).Length == 0)
+            {
+                // Dismiss the session
+                return null;
+            }
+
+            if (initialListOfItemsToBeIncluded.Count == 0)
+            {
+                return HandleAllItemsFilteredOut(reason, data.SelectedFilters, selectedFilters.Any(), completionRules);
+            }
+
+            var options = document?.Project?.Solution?.Options;
+            var highlightMatchingPortions = options?.GetOption(CompletionOptions.HighlightMatchingPortionsOfCompletionListItems, document.Project.Language) ?? true;
+            var showCompletionItemFilters = options?.GetOption(CompletionOptions.ShowCompletionItemFilters, document.Project.Language) ?? true;
+
+            var updatedFilters = showCompletionItemFilters
+                ? GetUpdatedFilters(initialListOfItemsToBeIncluded, data.SelectedFilters)
+                : ImmutableArray<AsyncCompletionData.CompletionFilterWithState>.Empty;
+
+            var highlightedList = GetHighlightedList(initialListOfItemsToBeIncluded, filterText, highlightMatchingPortions).ToImmutableArray();
+
+            // If this was deletion, then we control the entire behavior of deletion ourselves.
+            if (initialRoslynTriggerKind == CompletionTriggerKind.Deletion)
+            {
+                return HandleDeletionTrigger(initialListOfItemsToBeIncluded, filterText, updatedFilters, highlightedList);
+            }
+
+            Func<ImmutableArray<RoslynCompletionItem>, string, ImmutableArray<RoslynCompletionItem>> filterMethod;
+            if (completionService == null)
```

Comment:
how can we not have a completion service?

Reply:
Here is a scenario we just observed. In a Razor code, one types '@' and presses CTRL+J. Razor provides items to Editor. Roslyn either not called o provide items or responded that it could not provide any item.

Then, Editor decides to call Roslyn to filter items. All items were created by Razor. None of other has a reference to completion service.

---
In reply to: [232509548](https://github.com/dotnet/roslyn/pull/29016#discussion_r232509548) [](ancestors = 232509548)